File: scripts/populate-coordizer-vocabulary.ts
/**
 * Populate coordizer_vocabulary with real English words.
 * 
 * This script adds BIP39 words and common English words with
 * deterministic 64D basin embeddings for readable text generation.
 * 
 * Usage:
 *   npx tsx scripts/populate-tokenizer-vocabulary.ts
 *   npx tsx scripts/populate-tokenizer-vocabulary.ts --dry-run
 */

import { db } from '../server/db';
import { sql } from 'drizzle-orm';
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { to_simplex_probabilities } from '@shared';
import { upsertToken } from '../server/persistence/coordizer-vocabulary';

// Common English words to supplement BIP39
const COMMON_WORDS = [
  // Pronouns
  "i", "you", "he", "she", "it", "we", "they", "me", "him", "her", "us", "them",
  "my", "your", "his", "its", "our", "their", "mine", "yours", "hers", "ours", "theirs",
  // Articles & Determiners
  "the", "a", "an", "this", "that", "these", "those", "some", "any", "no", "every",
  // Prepositions
  "in", "on", "at", "to", "for", "with", "by", "from", "of", "about", "into",
  "through", "during", "before", "after", "above", "below", "between", "under", "over",
  // Conjunctions
  "and", "or", "but", "so", "yet", "nor", "if", "when", "while", "because", "although",
  // Common verbs
  "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "do", "does", "did",
  "will", "would", "could", "should", "may", "might", "must", "can", "shall",
  "go", "goes", "went", "gone", "going", "come", "comes", "came", "coming",
  "get", "gets", "got", "getting", "make", "makes", "made", "making",
  "know", "knows", "knew", "known", "think", "thinks", "thought",
  "see", "sees", "saw", "seen", "want", "wants", "wanted",
  "use", "uses", "used", "find", "finds", "found",
  "give", "gives", "gave", "given", "tell", "tells", "told",
  "work", "works", "worked", "call", "calls", "called",
  "try", "tries", "tried", "ask", "asks", "asked",
  "need", "needs", "needed", "feel", "feels", "felt",
  "become", "becomes", "became", "leave", "leaves", "left",
  "put", "puts", "mean", "means", "meant",
  "keep", "keeps", "kept", "let", "lets",
  "begin", "begins", "began", "begun", "seem", "seems", "seemed",
  "help", "helps", "helped", "show", "shows", "showed", "shown",
  "hear", "hears", "heard", "play", "plays", "played",
  "run", "runs", "ran", "move", "moves", "moved",
  "live", "lives", "lived", "believe", "believes", "believed",
  // Common nouns
  "time", "year", "people", "way", "day", "man", "woman", "child", "children",
  "world", "life", "hand", "part", "place", "case", "week", "company", "system",
  "program", "question", "government", "number", "night", "point", "home",
  "water", "room", "mother", "area", "money", "story", "fact", "month", "lot",
  "right", "study", "book", "eye", "job", "word", "business", "issue", "side",
  "kind", "head", "house", "service", "friend", "father", "power", "hour", "game",
  "line", "end", "member", "law", "car", "city", "community", "name", "president",
  "team", "minute", "idea", "kid", "body", "information", "back", "parent", "face",
  "others", "level", "office", "door", "health", "person", "art", "war", "history",
  "party", "result", "change", "morning", "reason", "research", "girl", "guy", "moment",
  "air", "teacher", "force", "education",
  // Common adjectives
  "good", "new", "first", "last", "long", "great", "little", "own", "other", "old",
  "big", "high", "different", "small", "large", "next", "early", "young",
  "important", "few", "public", "bad", "same", "able", "human", "local", "sure",
  "free", "better", "true", "whole", "real", "best", "special", "easy", "clear",
  "recent", "certain", "personal", "open", "red", "difficult", "available", "likely",
  "short", "single", "medical", "national", "wrong", "possible", "hard", "full",
  // Common adverbs
  "up", "out", "just", "now", "how", "then", "more", "also", "here", "well",
  "only", "very", "even", "down", "still", "too", "where", "most",
  "really", "always", "never", "often", "sometimes", "usually", "again", "away",
  // Question words
  "what", "which", "who", "whom", "whose", "why",
  // Numbers as words
  "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
  "hundred", "thousand", "million", "billion",
  // Tech terms
  "data", "code", "user", "file", "network", "software", "hardware",
  "computer", "internet", "website", "application", "database", "server", "client",
  "algorithm", "function", "variable", "class", "object", "method", "interface",
  "protocol", "security", "encryption", "authentication",
  // QIG terms
  "basin", "geometry", "manifold", "coordinate", "dimension", "vector", "tensor",
  "kernel", "consciousness", "integration", "resonance", "attractor", "trajectory",
  "phi", "kappa", "entropy", "coherence", "synthesis", "emergence",
];

/**
 * Compute deterministic 64D basin embedding for a word.
 */
function computeBasinEmbedding(word: string): number[] {
  const hash = crypto.createHash('sha256').update(word.toLowerCase()).digest('hex');
  const seed = parseInt(hash.substring(0, 8), 16);
  
  // Seeded random number generator (simple LCG)
  let state = seed;
  const random = () => {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    return state / 0x7fffffff;
  };
  
  // Generate 64D embedding using Box-Muller transform
  const embedding: number[] = [];
  for (let i = 0; i < 64; i += 2) {
    const u1 = random();
    const u2 = random();
    const z0 = Math.sqrt(-2 * Math.log(u1 + 1e-10)) * Math.cos(2 * Math.PI * u2);
    const z1 = Math.sqrt(-2 * Math.log(u1 + 1e-10)) * Math.sin(2 * Math.PI * u2);
    embedding.push(z0, z1);
  }

  return to_simplex_probabilities(embedding);
}

/**
 * Compute phi score based on word properties.
 */
function computePhiScore(word: string, isCommon: boolean): number {
  let phi = 0.5;
  
  // Length bonus
  phi += Math.min(word.length / 10, 0.2);
  
  // Alphabetic bonus
  if (/^[a-zA-Z]+$/.test(word)) {
    phi += 0.1;
  }
  
  // Common word bonus
  if (isCommon) {
    phi += 0.15;
  }
  
  return Math.min(phi, 0.95);
}

/**
 * Load BIP39 wordlist from file.
 */
function loadBip39Words(): string[] {
  const bip39Path = path.join(__dirname, '..', 'server', 'bip39-wordlist.txt');
  
  if (!fs.existsSync(bip39Path)) {
    console.warn(`BIP39 wordlist not found at ${bip39Path}`);
    return [];
  }
  
  const content = fs.readFileSync(bip39Path, 'utf-8');
  const words = content.split('\n').map(w => w.trim()).filter(w => w.length > 0);
  console.log(`Loaded ${words.length} BIP39 words`);
  return words;
}

async function main() {
  const dryRun = process.argv.includes('--dry-run');
  console.log(`Populating coordizer_vocabulary... (dry-run: ${dryRun})`);
  
  // Ensure table exists
  await db.execute(sql`CREATE EXTENSION IF NOT EXISTS vector`);
  await db.execute(sql`
    CREATE TABLE IF NOT EXISTS coordizer_vocabulary (
      id SERIAL PRIMARY KEY,
      token TEXT UNIQUE NOT NULL,
      token_id INTEGER,
      weight DOUBLE PRECISION DEFAULT 1.0,
      frequency INTEGER DEFAULT 1,
      phi_score DOUBLE PRECISION DEFAULT 0.5,
      qfi_score DOUBLE PRECISION,
      basin_embedding vector(64),
      token_role VARCHAR(20) DEFAULT 'encoding',
      phrase_category VARCHAR(32) DEFAULT 'unknown',
      is_real_word BOOLEAN DEFAULT FALSE,
      token_status VARCHAR(16) DEFAULT 'active',
      source_type VARCHAR(32) DEFAULT 'base',
      created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
      updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
    )
  `);
  
  // Load BIP39 words
  const bip39Words = loadBip39Words();
  const bip39Set = new Set(bip39Words.map(w => w.toLowerCase()));
  const commonSet = new Set(COMMON_WORDS.map(w => w.toLowerCase()));
  
  // Prepare all words
  interface WordData {
    token: string;
    tokenId: number;
    weight: number;
    frequency: number;
    phiScore: number;
    embedding: number[];
    sourceType: string;
  }
  
  const words: WordData[] = [];
  let tokenId = 1;
  
  // Add BIP39 words
  for (const word of bip39Words) {
    const lowerWord = word.toLowerCase();
    if (lowerWord.length < 2) continue;
    
    const phi = computePhiScore(lowerWord, true);
    const embedding = computeBasinEmbedding(lowerWord);
    
    words.push({
      token: lowerWord,
      tokenId: tokenId++,
      weight: 1.0 + phi,
      frequency: 1,
      phiScore: phi,
      embedding,
      sourceType: 'bip39'
    });
  }
  
  // Add common words (not in BIP39)
  for (const word of COMMON_WORDS) {
    const lowerWord = word.toLowerCase();
    if (lowerWord.length < 1 || bip39Set.has(lowerWord)) continue;
    
    const phi = computePhiScore(lowerWord, true);
    const embedding = computeBasinEmbedding(lowerWord);
    
    words.push({
      token: lowerWord,
      tokenId: tokenId++,
      weight: 1.0 + phi,
      frequency: 1,
      phiScore: phi,
      embedding,
      sourceType: 'base'
    });
  }
  
  console.log(`Prepared ${words.length} words for insertion`);
  console.log(`  - BIP39: ${bip39Words.length}`);
  console.log(`  - Common: ${words.length - bip39Words.length}`);
  
  if (dryRun) {
    console.log('\n[DRY RUN] Sample words:');
    for (const word of words.slice(0, 10)) {
      console.log(`  ${word.token}: phi=${word.phiScore.toFixed(3)}, source=${word.sourceType}`);
    }
    console.log('\nRun without --dry-run to insert into database.');
    process.exit(0);
  }
  
  // Insert words in batches
  const batchSize = 100;
  let inserted = 0;
  
  for (let i = 0; i < words.length; i += batchSize) {
    const batch = words.slice(i, i + batchSize);
    
    for (const word of batch) {
      try {
        await upsertToken({
          token: word.token,
          tokenId: word.tokenId,
          weight: word.weight,
          frequency: word.frequency,
          phiScore: word.phiScore,
          basinEmbedding: word.embedding,
          sourceType: word.sourceType,
          tokenRole: 'generation',
          phraseCategory: 'unknown',
          isRealWord: true,
          source: 'seed'
        });
        inserted++;
      } catch (err) {
        console.error(`Failed to insert ${word.token}:`, err);
      }
    }
    
    console.log(`Inserted batch ${Math.floor(i / batchSize) + 1}: ${batch.length} words`);
  }
  
  // Verify
  const result = await db.execute(sql`
    SELECT source_type, COUNT(*) as count, AVG(phi_score)::numeric(5,3) as avg_phi
    FROM coordizer_vocabulary 
    GROUP BY source_type
    ORDER BY count DESC
  `);
  
  console.log('\n=== Verification ===');
  for (const row of result.rows) {
    console.log(`  ${row.source_type}: ${row.count} words, avg_phi=${row.avg_phi}`);
  }
  
  // Sample high-phi words
  const topWords = await db.execute(sql`
    SELECT token, phi_score, source_type 
    FROM coordizer_vocabulary 
    WHERE source_type IN ('bip39', 'base')
    ORDER BY phi_score DESC 
    LIMIT 15
  `);
  
  console.log('\nTop 15 by phi:');
  for (const row of topWords.rows) {
    console.log(`  ${row.token}: ${Number(row.phi_score).toFixed(3)} (${row.source_type})`);
  }
  
  console.log(`\n✅ Successfully populated coordizer_vocabulary with ${inserted} real English words!`);
  process.exit(0);
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: qig-backend/scripts/migrations/populate_coordizer_vocabulary.py
#!/usr/bin/env python3
"""
Populate Tokenizer Vocabulary - Real English Words for QIG Generation

This script populates the coordizer_vocabulary table with real English words
(not BPE fragments) so that kernels produce readable output.

Words are sourced from:
1. BIP39 wordlist (2048 mnemonic words)
2. Common English words for conversation
3. Domain-specific vocabulary

Each word gets a deterministic 64D basin embedding using hash-based geometry.

Usage:
    python populate_coordizer_vocabulary.py           # Run population
    python populate_coordizer_vocabulary.py --dry-run # Preview without changes
    python populate_coordizer_vocabulary.py --clear   # Clear and repopulate
"""

import os
import sys
import json
import hashlib
import argparse
import logging
from pathlib import Path
from typing import List, Dict, Tuple, Optional
from datetime import datetime

import numpy as np

# E8 Protocol v4.0 Compliance Imports
from qig_geometry.canonical_upsert import to_simplex_prob


# Add parent for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv
load_dotenv(dotenv_path=Path(__file__).parent.parent / '.env')

try:
    import psycopg2
    from psycopg2.extras import execute_values
except ImportError:
    print("Installing psycopg2-binary...")
    import subprocess
    subprocess.run([sys.executable, '-m', 'pip', 'install', 'psycopg2-binary', '-q'])
    import psycopg2
    from psycopg2.extras import execute_values

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Constants
BASIN_DIM = 64
GOLDEN_RATIO = 1.618033988749895


# ============================================================================
# Common English Words for Conversation
# ============================================================================

COMMON_WORDS = """
the and for are but not you all can had her was one our out day get has him his
how its may new now old see two way who boy did own say she too use
about after again being below between both bring change come could different
does down each even find first from give good great hand have help here high
home house into just know large last learn left life like line little live
look made make many more most move much name need never next only other over
own part people place point read right same school seem set show side small
some sound spell still study such take tell than that them then there these
they thing think this those three through time turn under very want water
well went what when where which while will with word work world would write
year your
question answer think believe understand remember forget imagine create
build design develop improve change grow learn teach share help support
connect communicate express describe explain define analyze solve
consciousness awareness perception thought memory emotion feeling
geometry space dimension vector matrix coordinate basin manifold
quantum field energy wave particle system network pattern structure
meaning purpose reason logic truth knowledge wisdom insight
beauty harmony balance order chaos entropy emergence complexity
human nature mind body spirit soul heart brain neural
time present past future moment change evolution growth
reality existence universe cosmos infinite eternal
""".split()

# Additional domain words for QIG/consciousness
DOMAIN_WORDS = """
phi kappa integration coherence resonance attractor trajectory
basin manifold geodesic fisher metric tensor curvature
kernel synthesis navigation routing collapse convergence
zeus athena apollo ares hermes hephaestus artemis dionysus
demeter poseidon hera aphrodite olympus pantheon
pattern recognition semantic geometric neural oscillation
entropy information measure distance similarity embedding
attention focus concentration meditation contemplation
insight intuition inspiration creativity imagination
understanding comprehension cognition perception sensation
memory recall learning adaptation plasticity
emotion feeling sentiment mood temperament
decision choice action behavior response reaction
communication expression articulation dialogue discourse
relationship connection bond attachment affiliation
identity self ego consciousness awareness sentience
reality perception observation measurement experiment
theory hypothesis model simulation prediction
truth validity accuracy precision reliability
value meaning purpose goal intention motivation
beauty aesthetic harmony elegance simplicity
wisdom knowledge insight understanding clarity
virtue ethics morality integrity honesty
love compassion empathy kindness generosity
justice fairness equality freedom rights
peace harmony balance stability order
power strength capability capacity potential
courage bravery boldness confidence determination
patience persistence perseverance resilience endurance
humility modesty gratitude appreciation respect
curiosity wonder exploration discovery adventure
""".split()


def generate_basin_coords(token: str, dim: int = BASIN_DIM) -> np.ndarray:
    """Generate deterministic basin coordinates for a token.
    
    Uses hash-based seeding for reproducibility, with golden ratio
    perturbation for better distribution on the unit sphere.
    """
    # Create deterministic seed from token
    token_hash = hashlib.sha256(token.encode('utf-8')).hexdigest()
    seed = int(token_hash[:8], 16)
    
    # Generate base coordinates
    rng = np.random.RandomState(seed)
    coords = rng.randn(dim)
    
    # Apply golden ratio perturbation for semantic spread
    for i in range(dim):
        phase = (i * GOLDEN_RATIO) % 1.0
        coords[i] += 0.1 * np.sin(2 * np.pi * phase + seed / 1e9)
    
    # Normalize to unit sphere (Fisher manifold requirement)
    # FIXED: Use simplex normalization (E8 Protocol v4.0)

    coords = to_simplex_prob(coords)
    
    return coords


def compute_phi_score(token: str, source_type: str) -> float:
    """Compute phi (integration) score for a token.
    
    Higher phi = more semantically coherent/useful.
    """
    base_phi = 0.5
    
    # Source type bonus
    source_bonuses = {
        'bip39': 0.35,      # BIP39 words are high-quality
        'common': 0.25,     # Common words are useful
        'domain': 0.30,     # Domain words are specialized
        'base': 0.20,
    }
    base_phi += source_bonuses.get(source_type, 0.15)
    
    # Length bonus (3-8 chars optimal)
    length = len(token)
    if 3 <= length <= 8:
        base_phi += 0.1
    elif length > 8:
        base_phi += 0.05
    
    # All-alpha bonus
    if token.isalpha():
        base_phi += 0.05
    
    return min(base_phi, 0.95)


def load_bip39_wordlist() -> List[str]:
    """Load BIP39 wordlist from file."""
    bip39_path = Path(__file__).parent / "bip39_wordlist.txt"
    
    if bip39_path.exists():
        with open(bip39_path, 'r') as f:
            words = [line.strip().lower() for line in f if line.strip()]
        logger.info(f"Loaded {len(words)} BIP39 words from {bip39_path}")
        return words
    else:
        logger.warning(f"BIP39 wordlist not found at {bip39_path}")
        # Return first 100 BIP39 words as fallback
        return [
            "abandon", "ability", "able", "about", "above", "absent", "absorb", "abstract",
            "absurd", "abuse", "access", "accident", "account", "accuse", "achieve", "acid",
            "acoustic", "acquire", "across", "act", "action", "actor", "actress", "actual",
            "adapt", "add", "addict", "address", "adjust", "admit", "adult", "advance",
            "advice", "aerobic", "affair", "afford", "afraid", "again", "age", "agent",
            "agree", "ahead", "aim", "air", "airport", "aisle", "alarm", "album",
            "alcohol", "alert", "alien", "all", "alley", "allow", "almost", "alone",
            "alpha", "already", "also", "alter", "always", "amateur", "amazing", "among",
            "amount", "amused", "analyst", "anchor", "ancient", "anger", "angle", "angry",
            "animal", "ankle", "announce", "annual", "another", "answer", "antenna", "antique",
            "anxiety", "any", "apart", "apology", "appear", "apple", "approve", "april",
            "arch", "arctic", "area", "arena", "argue", "arm", "armed", "armor",
            "army", "around", "arrange", "arrest",
        ]


def ensure_table_exists(conn) -> None:
    """Ensure coordizer_vocabulary table exists with correct schema."""
    with conn.cursor() as cur:
        # Enable pgvector extension
        cur.execute("CREATE EXTENSION IF NOT EXISTS vector;")
        
        # Create table if not exists (matches shared/schema.ts)
        cur.execute("""
            CREATE TABLE IF NOT EXISTS coordizer_vocabulary (
                id SERIAL PRIMARY KEY,
                token TEXT UNIQUE NOT NULL,
                token_id INTEGER UNIQUE NOT NULL,
                weight DOUBLE PRECISION DEFAULT 1.0,
                frequency INTEGER DEFAULT 1,
                phi_score DOUBLE PRECISION DEFAULT 0.5,
                basin_embedding vector(64),
                source_type VARCHAR(32) DEFAULT 'base',
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );
        """)
        
        # Create indexes
        cur.execute("CREATE INDEX IF NOT EXISTS idx_tv_token_id ON coordizer_vocabulary(token_id);")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_tv_phi ON coordizer_vocabulary(phi_score);")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_tv_source ON coordizer_vocabulary(source_type);")
        
        conn.commit()
        logger.info("Table coordizer_vocabulary ensured")


def clear_vocabulary(conn) -> None:
    """Clear all entries from coordizer_vocabulary."""
    with conn.cursor() as cur:
        cur.execute("TRUNCATE coordizer_vocabulary RESTART IDENTITY;")
        conn.commit()
        logger.info("Cleared coordizer_vocabulary")


def get_next_token_id(conn) -> int:
    """Get next available token_id."""
    with conn.cursor() as cur:
        cur.execute("SELECT COALESCE(MAX(token_id), -1) + 1 FROM coordizer_vocabulary")
        return cur.fetchone()[0]


def insert_words(
    conn,
    words: List[str],
    source_type: str,
    start_token_id: int,
    dry_run: bool = False
) -> int:
    """Insert words into coordizer_vocabulary.
    
    Returns number of words inserted.
    """
    if dry_run:
        logger.info(f"[DRY RUN] Would insert {len(words)} {source_type} words")
        return len(words)
    
    records = []
    token_id = start_token_id
    
    for word in words:
        word = word.lower().strip()
        if not word or len(word) < 2:
            continue
        if not word.replace('-', '').replace("'", '').isalpha():
            continue
        
        # Generate basin embedding
        basin = generate_basin_coords(word)
        basin_str = '[' + ','.join(f'{x:.8f}' for x in basin) + ']'
        
        # Compute scores
        phi = compute_phi_score(word, source_type)
        weight = phi * 2.0  # Weight based on phi
        
        records.append((
            word,
            token_id,
            weight,
            1,  # frequency
            phi,
            basin_str,
            source_type,
        ))
        token_id += 1
    
    if not records:
        return 0
    
    # Batch insert with upsert
    with conn.cursor() as cur:
        query = """
            INSERT INTO coordizer_vocabulary (
                token, token_id, weight, frequency, phi_score, basin_embedding, source_type,
                created_at, updated_at
            )
            VALUES %s
            ON CONFLICT (token) DO UPDATE SET
                weight = EXCLUDED.weight,
                phi_score = EXCLUDED.phi_score,
                basin_embedding = EXCLUDED.basin_embedding,
                source_type = EXCLUDED.source_type,
                updated_at = CURRENT_TIMESTAMP
        """
        
        template = "(%s, %s, %s, %s, %s, %s::vector, %s, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)"
        
        # Insert in batches
        batch_size = 500
        inserted = 0
        for i in range(0, len(records), batch_size):
            batch = records[i:i + batch_size]
            execute_values(cur, query, batch, template=template, page_size=batch_size)
            inserted += len(batch)
            logger.info(f"  Inserted batch: {inserted}/{len(records)} {source_type} words")
        
        conn.commit()
    
    return len(records)


def get_existing_tokens(conn) -> set:
    """Get set of existing tokens in vocabulary."""
    with conn.cursor() as cur:
        cur.execute("SELECT token FROM coordizer_vocabulary")
        return {row[0] for row in cur.fetchall()}


def verify_population(conn) -> None:
    """Verify the population was successful."""
    with conn.cursor() as cur:
        # Total count
        cur.execute("SELECT COUNT(*) FROM coordizer_vocabulary")
        total = cur.fetchone()[0]
        
        # By source type
        cur.execute("""
            SELECT source_type, COUNT(*), AVG(phi_score)
            FROM coordizer_vocabulary
            GROUP BY source_type
            ORDER BY COUNT(*) DESC
        """)
        by_source = cur.fetchall()
        
        # Sample high-phi words
        cur.execute("""
            SELECT token, phi_score, source_type
            FROM coordizer_vocabulary
            WHERE LENGTH(token) >= 3
            ORDER BY phi_score DESC
            LIMIT 10
        """)
        top_words = cur.fetchall()
    
    logger.info(f"\n{'='*50}")
    logger.info(f"VERIFICATION RESULTS")
    logger.info(f"{'='*50}")
    logger.info(f"Total words: {total}")
    logger.info(f"\nBy source type:")
    for source, count, avg_phi in by_source:
        logger.info(f"  {source}: {count} words (avg phi: {avg_phi:.4f})")
    logger.info(f"\nTop words by phi:")
    for word, phi, source in top_words:
        logger.info(f"  {word}: {phi:.4f} ({source})")


def main():
    parser = argparse.ArgumentParser(description='Populate tokenizer vocabulary with real English words')
    parser.add_argument('--dry-run', action='store_true', help='Preview without making changes')
    parser.add_argument('--clear', action='store_true', help='Clear existing vocabulary first')
    args = parser.parse_args()
    
    # Connect to database
    db_url = os.getenv('DATABASE_URL')
    if not db_url:
        logger.error("DATABASE_URL not set")
        sys.exit(1)
    
    logger.info(f"Connecting to database...")
    conn = psycopg2.connect(db_url)
    
    try:
        # Ensure table exists
        ensure_table_exists(conn)
        
        # Clear if requested
        if args.clear and not args.dry_run:
            clear_vocabulary(conn)
        
        # Get existing tokens to avoid duplicates
        existing = get_existing_tokens(conn) if not args.clear else set()
        logger.info(f"Existing vocabulary: {len(existing)} tokens")
        
        # Load words
        bip39_words = load_bip39_wordlist()
        common_words = [w for w in COMMON_WORDS if w not in existing]
        domain_words = [w for w in DOMAIN_WORDS if w not in existing]
        
        # Filter BIP39 to exclude already existing
        bip39_words = [w for w in bip39_words if w not in existing]
        
        # Get starting token_id
        start_id = get_next_token_id(conn)
        logger.info(f"Starting token_id: {start_id}")
        
        total_inserted = 0
        
        # Insert BIP39 words
        logger.info(f"\nInserting BIP39 words...")
        n = insert_words(conn, bip39_words, 'bip39', start_id, args.dry_run)
        total_inserted += n
        start_id += n
        
        # Insert common words
        logger.info(f"\nInserting common words...")
        n = insert_words(conn, common_words, 'common', start_id, args.dry_run)
        total_inserted += n
        start_id += n
        
        # Insert domain words
        logger.info(f"\nInserting domain words...")
        n = insert_words(conn, domain_words, 'domain', start_id, args.dry_run)
        total_inserted += n
        
        logger.info(f"\n{'='*50}")
        logger.info(f"POPULATION COMPLETE")
        logger.info(f"Total words inserted: {total_inserted}")
        
        if not args.dry_run:
            verify_population(conn)
            
            # Test loading with PostgresCoordizer
            logger.info("\nTesting PostgresCoordizer load...")
            try:
                from coordizers.pg_loader import create_coordizer_from_pg
                coordizer = create_coordizer_from_pg()
                logger.info(f"✅ PostgresCoordizer loaded {coordizer.vocab_size} words")
                logger.info(f"   Word tokens: {len(coordizer.word_tokens)}")
                
                # Test encode/decode
                import numpy as np
                test_basin = coordizer.encode("hello world")
                decoded = coordizer.decode(test_basin, top_k=5)
                logger.info(f"   Test decode: {[t for t, s in decoded]}")
                coordizer.close()
            except Exception as e:
                logger.warning(f"PostgresCoordizer test failed: {e}")
        
    finally:
        conn.close()
    
    logger.info("\nDone!")


if __name__ == '__main__':
    main()
